mitodl/open-discussions
# Allow CRUD operations for UserLists via API

#### Pre-Flight checklist

#### What are the relevant tickets?
Closes #2321 

#### What's this PR do?
- Allows UserLists to be created/modified/deleted via REST API calls
- Allows UserListItems to be created/modified/deleted via REST API calls
- Adds permissions for the above

#### How should this be manually tested?
- Go to `http://od.odl.local:8063/api/v0/userlists/` and try to post the following JSON, it should create a new list with correct field values:
```
{
    "title": "MY NEW LIST",
    "short_description": "A description of my new list",
    "privacy_level": "private",
    "list_type": "Learning Path"
}
```

- Go to `http://od.odl.local:8063/api/v0/userlists/{list_id}` and PATCH the following data, the UserList title should be updated:
```
{
"title": "My new title"
}
```
- Try the above again but with a list_id that belongs to another user, you should not be allowed to update it.

- Go to `http://od.odl.local:8063/api/v0/userlists/{id}` and PATCH the following data, two list items should be added to your list:
```
{ "items": 
  [
    {
         "content_type": "course",
        "object_id": <id of existing course>
    },
    {
         "content_type": "program",
        "object_id": <id of existing program>
    }
  ]
}
```
- Try the above again but with a non-existent course id, you should get an error
- PATCH the following data, the existing items' positions should be updated
```
{ "items": [
    {
         "id": <id of item 1>,
         "position": 33
    },
    {
         "id": <id of item 2>,
         "position": 34
    }
]
}
```
- PATCH the following data, the item should be deleted.
```
{ "items": [
    {
         "id": <id of item 1>,
         "delete": true
    }
]
}
```

- PATCH the following data, the 2nd item should be deleted.
```
{ "items": [
    {
         "object_id": <object_id of item 1>,
         "content_type": "course",
         "delete": true
    }
]
}
```
